### PR TITLE
Support saving decode cache

### LINKED ISSUE #129: Support saving KV cache after decoding
Currently, KV cache is saved after prefilling. Saving KV cache after decoding may be advantageous for applications such as multi-turn conversation.

### PR DESCRIPTION
**Pull request by YaoJiayi**
This pr resolves https://github.com/LMCache/LMCache/issues/129.
Also related to LMCache PR https://github.com/LMCache/LMCache/pull/149.

### CODE REVIEW

File: lmcache_vllm/vllm_adapter.py [hunk 1]
```diff
@@ -112,7 +113,7 @@ def lmcache_should_retrieve(
 
 def lmcache_should_store(
         model_input: "ModelInputForGPUWithSamplingMetadata", 
-        kv_caches: List[torch.Tensor]) -> bool:
+        kv_caches: List[torch.Tensor]) -> str:
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_adapter.py]: Maybe use ENUM class instead of str.

File: lmcache_vllm/vllm_injection.py [hunk 2]
```diff
@@ -163,6 +172,57 @@ def new_execute_model(
...
+    res: List[int],
+) -> List[int]:
+    """
+    remove empty tokens to enable caching decode KV cache
+    """
+    mistral_family = ['mistralai/Mistral-7B-Instruct-v0.2']
+    if tokenizer_id in mistral_family:
+        empty_token = 28705
+        res = [x for x in res if x != empty_token]
+    return res
```
Comment (KuntaiDu) [inline lmcache_vllm/vllm_injection.py:187]: I am not fully understanding why we should exclude empty token here. Can you elaborate?

Comment (YaoJiayi) [reply lmcache_vllm/vllm_injection.py:187]: In the first run, text1 produces answer1. The hash is based on (text1, answer1). 
In the second run, when you pass (text1, answer1) as the input, there will be a padding token between them. The hash is based on (text1, pad_tok, answer1).

Comment (XbzOnGit) [reply lmcache_vllm/vllm_injection.py:187]: But that token "▁": 28705 appears even in the first prompt(not relevant with decoding caching) for many times, according to the loggings of vllm. Is it correct to remove them?

File: lmcache_vllm/vllm_adapter.py [hunk 3]
```diff
@@ -123,9 +130,10 @@ def lmcache_should_store(
 
     :return: True if we should store KV into LMCache, False otherwise.
```
Comment (ApostaC) [inline lmcache_vllm/vllm_adapter.py]: Can you change the docstring here?
What's the meaning of PREFILL, DECODE, and NONE in `StoreStatus`?

File: lmcache_vllm/vllm_injection.py (hunk 2 above)
Comment (ApostaC) [reply lmcache_vllm/vllm_injection.py:187]: Is this only for the mistral model? Or different models may have different padding tokens?
If it's the latter case, we may need a more systematic solution (for instance, save the text and newly generated tokens temporarily and do some matching)

File: lmcache_vllm/vllm_adapter.py [hunk 4]
```diff
@@ -192,22 +217,46 @@ def lmcache_store_kv(
...
         keys, values = [], []
         kv_tuple_list = []
 
+        if is_prefill:
+            start_pos = sum(seq_lens[:idx])
+            end_pos = start_pos + slen
+            current_tokens = input_tokens_tensor[start_pos:end_pos]
+            current_slot_mapping = slot_mapping_flat[start_pos:end_pos]
+        else:
+            if slen % engine.chunk_size != 0:
```
Comment (ApostaC) [inline lmcache_vllm/vllm_adapter.py:231]: @YaoJiayi Let's find some time to discuss this maybe? I understand this is a temporary solution, but we need to think about if there is a better way to do it.

**Review by ApostaC — COMMENTED** [review]
Thanks, Jiayi! Got some comments.

File: lmcache_vllm/vllm_injection.py (hunk 2 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_injection.py:187]: Someone need to check whether this patch is needed for other models, which I think is a "good first issue".
I do agree we should have a more systematic solution. Maybe some kind of non-exact matching algorithm is needed.

File: lmcache_vllm/vllm_adapter.py (hunk 4 above)
Comment (YaoJiayi) [reply lmcache_vllm/vllm_adapter.py:231]: Agreed

File: lmcache_vllm/vllm_injection.py (hunk 2 above)
Comment (ApostaC) [reply lmcache_vllm/vllm_injection.py:187]: Do we have a ticket for this issue? If so, I'm good with the current code.

**Review by ApostaC — APPROVED** [review]
LGTM. @YaoJiayi Don't forget to create a new ticket for the `remove_empty_token` related issue if you haven't done it yet.

File: lmcache_vllm/vllm_injection.py (hunk 2 above)
Comment (KuntaiDu) [reply lmcache_vllm/vllm_injection.py:187]: Agree with Yihua.

**Pull request merged**